src/oph/ehoks/palaute/opiskelija.clj

```diff
@@ -160,3 +167,74 @@
                          kysely-type
                          opts))
                    hoksit))))
+
+(defn create-arvo-kyselylinkki!
+  "For the given palaute, make Arvo call for creating its kyselylinkki
+  and return the Arvo reply."
+  [palaute]
+  (-> palaute
+      (select-keys [:hoks-id :kyselytyyppi])
+      (->> (palaute/get-for-arvo-by-hoks-id-and-kyselytyyppi! db/spec))
+      (or (throw (ex-info "No Arvo-processable palaute found"
+                          {:palaute palaute})))
+      (utils/to-underscore-keys)
+      ;; FIXME: add up-to-date values from Koski here
+      (assoc :metatiedot {:tila "ei_lahetetty"}
+             :request_id (str (UUID/randomUUID)))
+      (#(assoc % :tutkinnonosat_hankkimistavoittain
+               {:oppisopimus (:tutkinnonosat_oppisopimus %)
+                :koulutussopimus (:tutkinnonosat_koulutussopimus %)
+                :oppilaitosmuotoinenkoulutus
+                (:tutkinnonosat_oppilaitosmuotoinenkoulutus %)}))
+      (update :tutkinnon_suorituskieli #(or % "fi"))
+      (update :koulutustoimija_oid #(or % ""))
+      (update :tutkintotunnus str)
+      (update :kyselyn_tyyppi translate-kyselytyyppi)
+      (dissoc :tila :tutkinnonosat_koulutussopimus :tutkinnonosat_oppisopimus
+              :tutkinnonosat_oppilaitosmuotoinenkoulutus)
+      (arvo/create-kyselytunnus!)))
+
+(defn create-and-save-arvo-kyselylinkki!
+  "Update given palaute with a kyselylinkki from Arvo."
+  [palaute]
+  (try
+    (jdbc/with-db-transaction
+      [tx db/spec]
+      (let [response (create-arvo-kyselylinkki! palaute)
+            for-db (rename-keys response {:kysely_linkki :url})]
+        (try
+          (palaute/save-arvo-tunniste! tx palaute for-db
+                                       {:arvo_response response})
+          (catch Exception e
+            (log/error "error while saving arvo tunniste" (:tunnus response)
+                       "; trying to delete kyselylinkki")
+            ;; FIXME: create chained exception if this throws
+            (arvo/delete-kyselytunnus (:tunnus response))
+            (log/info "successfully deleted kyselylinkki" (:tunnus response))
+            (throw e)))))
+    (catch Exception e
+      (log/error e "while processing palaute" palaute)
+      (palautetapahtuma/insert!
+        db/spec
+        {:palaute-id      (:id palaute)
+         :vanha-tila      (:tila palaute)
+         :uusi-tila       (:tila palaute)
+         :tapahtumatyyppi "arvo_luonti"
+         :syy             "arvo_kutsu_epaonnistui"
+         :lisatiedot      {:errormsg (.getMessage e)
+                           :body (:body (ex-data e))}})
+      (throw e)))
+  (dynamodb/sync-amis-herate! (:hoks-id palaute) (:kyselytyyppi palaute)))
```

Comment:
Ehdotus: Näiden siirtoa voi harkita olemassa olevaan `palaute.opiskelija.kyselylinkki`-nimiavaruuteen, jolloin olisi varmaan `kyselylinkki/create-and-save!`. Omasta mielestäni jokaisessa funktion nimessä ei ole pakko mainita sanaa "arvo", koska kontekstista on pääteltävissä, että Arvoon varmastikin tehdään kutsuja osana tuota noita funktiota.

Reply:
Jeps, tää funktio tosin _myös_ pitäis [yhdistää](https://jira.eduuni.fi/browse/EH-1737) työelämäpalautteen vastaavaan.  Nää nimiavaruudet on vähän ongelmallisesti jaettu siinä mielessä, että ne ohjaa jaottelemaan koodia amis- ja tep-palautteen välillä, kun tarkoitus olis jakaa koodia mahdollisimman paljon.

Mut mä katson tuossa EH-1737:ssa, mikä sen nimiavaruuden nimi kannattais olla, niin ei välttämättä tuu enää tätä jaottelua.

Reply:
Aivan juu. Itse oon noiden nimiavaruuksien suhteen ajatellut niin, että amis- ja tep-palautteille yhteinen koodi eläisi tuolla `oph.ehoks.palaute`-nimiavaruudessa, ja ainoastaan tapauksissa jossa koodia ei ole mahdollista jakaa koodi menisi noihin `opiskelija`- ja `tyoelama`-nimiavaruuksiin. Olisiko kyselylinkkeihin liittyvälle jaetulle koodilla sopiva paikka nimiavaruus `oph.ehoks.palaute.kyselylinkki`?